elastic/fleet-server
# Allow multiple ES outputs as long as they are the same ES

## What is the problem this PR solves?

It fixes #1672. Currently fleet-server does not handle correctly the multiple ES outputs the fleet-ui might create.

## How does this PR solve the problem?

It creates a non-breaking change on the agent documents to store ES API key information per output.

## How to test this PR locally

- add a second ES output:
    - go to Fleet > Settings
    - outputs > add output
    - just copy the data from the `default` output and git it another name. Let's call it `non-default`
- create/edit a policy to use `non-default` output for integrations. Let the monitoring using the default:
    - navigate to Fleet > Agent policies > your_policy
    - go to the settings "tab"
    - select `non-default` as the output for _Output for integrations_. (You'll need a paid license or use Elastic Cloud)
- enroll a Elastic Agent on this policy with some integration

At first both outputs should have their on API keys. Use [`elastic-agent inspect`](https://www.elastic.co/guide/en/fleet/current/elastic-agent-cmd-options.html#elastic-agent-inspect-command) to check that.

Now modify the policy in a way that one of the outputs has its permissions changed, but not the other. I usually add and remove the Endpoint Security. Then both outputs should have the same API key, again, use the `inspect` command to check.

### [Here](https://gist.github.com/AndersonQ/b4dcfef8e9d203c848364ceda497daa5) you can see how I reproduced the issue and my investigation

## Checklist


## Author's checklist

- [x] Modify the agent model to support multiple ES outputs: add fields, don't change the existing ones, keep backward compatible
- [x] ensure new agent model is correctly saved and loaded (Marshal/UnMarshaled)
- [x] add migration process to internal/pkg/dl/migration.go
    - [x] add doc explaining the migration
- [x] when preparing ES output: 
    - [x] set the default API key to empty -> force all API keys to be regenerated
    - [x] set old fields to empty string/nil
    - [x] use new fields to store output API key data 
- [x] update painless script to use new agent model
- [x] ensure the `model.Agent` aren't used (except for tests checking they aren't populated anymore)
    - [x] update `handleAck` to use the new model
    - [x] update `monitor` to use the new model
- [x] on upgrade bump the coordinator index
- [x] ensure we invalidate the api keys from DefaultAPIKeyHistory (the migration should take care of it)
- [x] add output name to api key metadata
- [x] use only new agent model
    - [x] adjust `internal/pkg/coordinator/monitor_integration_test.go` to use new Agent model
    - [x] adjust `internal/pkg/coordinator/monitor.go` to use new Agent model
    - [x] ensure there is no reference to the deprecated agent fields. As the migration takes care of migrating every agent document, then it should be safe to remove the deprecated fields from the agent Schema
- [ ] **try/nice to have**: on agent checkin, after fleet-server upgrade to 8.4, try to search api key by metadata.agentId and invalidate the "old" ones. So far I don't think it's possible.
    - I think the most viable here would be to invalidate all the API keys during the migration


## Things to remember to test
 - ensure all API keys (access + outputs) are invalidated on Unenroll 
 - ensure that given an agent which the `default_api_key_history` field ins't empty, these keys get retired on ack and unenroll
 - ensure the existing agents documents get migrated (a.k.a the migration happens)
 - ensure the migrations is idempotent. Start the new fleet-server a few times and ensure it does not try to update the already updated agent documents. it can also be checked by looking at the logs. At the en of each migration a log like the following is produced.
log message: `migration MIGRATION_NAME done`
full log:
```
{"_index":".ds-logs-elastic_agent.fleet_server-default-2022.08.11-000001","_id":"sft_kYIB00fSKmPF2k05","_version":1,"_score":0,"_source":{"fleet.migration.retries.search":0,"agent":{"name":"fleet-server-dev","id":"308052ab-bca4-4eaa-8787-904feea56fa0","type":"filebeat","ephemeral_id":"b21f3b26-c6a3-4da6-9edf-56fc01813ebf","version":"8.5.0"},"log":{"file":{"path":"/opt/Elastic/Agent/data/elastic-agent-6d830e/logs/default/fleet-server-20220812-15.ndjson"},"offset":6005},"fleet.migration.version_conflicts":0,"elastic_agent":{"id":"308052ab-bca4-4eaa-8787-904feea56fa0","version":"8.5.0","snapshot":true},"fleet.migration.es.timed_out":false,"fleet.migration.total":1,"ecs":{"version":"8.0.0"},"fleet.migration.total.duration":27.307903,"fleet.migration.deleted":0,"fleet.migration.updated":1,"host":{"hostname":"fleet-server-dev","os":{"kernel":"5.13.0-52-generic","codename":"impish","name":"Ubuntu","type":"linux","family":"debian","version":"21.10 (Impish Indri)","platform":"ubuntu"},"containerized":false,"ip":["10.0.2.15","fe80::6c:bfff:fe6f:c40","192.168.56.43","fe80::a00:27ff:fed8:3139"],"name":"fleet-server-dev","id":"c287e219ca4f4643aa206fc914242e26","mac":["02:6c:bf:6f:0c:40","08:00:27:d8:31:39"],"architecture":"x86_64"},"log.level":"info","event":{"agent_id_status":"verified","ingested":"2022-08-12T10:00:51Z","dataset":"elastic_agent.fleet_server"},"fleet.migration.es.took":26,"fleet.migration.retries.bulk":0,"fleet.migration.name":"AgentOutputs","service.name":"fleet-server","fleet.migration.noops":0,"message":"migration AgentOutputs done","input":{"type":"filestream"},"@timestamp":"2022-08-12T10:00:43.961Z","data_stream":{"namespace":"default","type":"logs","dataset":"elastic_agent.fleet_server"},"fleet.migration.batches":1},"fields":{"fleet.migration.retries.search":[0],"elastic_agent.version":["8.5.0"],"host.os.name.text":["Ubuntu"],"host.hostname":["fleet-server-dev"],"host.mac":["02:6c:bf:6f:0c:40","08:00:27:d8:31:39"],"fleet.migration.total":[1],"host.os.version":["21.10 (Impish Indri)"],"fleet.migration.total.duration":[27.307903],"host.os.name":["Ubuntu"],"log.level":["info"],"agent.name":["fleet-server-dev"],"host.name":["fleet-server-dev"],"event.agent_id_status":["verified"],"fleet.migration.retries.bulk":[0],"fleet.migration.name":["AgentOutputs"],"host.os.type":["linux"],"fleet.migration.noops":[0],"input.type":["filestream"],"log.offset":[6005],"data_stream.type":["logs"],"host.architecture":["x86_64"],"agent.id":["308052ab-bca4-4eaa-8787-904feea56fa0"],"ecs.version":["8.0.0"],"host.containerized":[false],"agent.version":["8.5.0"],"host.os.family":["debian"],"fleet.migration.version_conflicts":[0],"fleet.migration.es.timed_out":[false],"host.ip":["10.0.2.15","fe80::6c:bfff:fe6f:c40","192.168.56.43","fe80::a00:27ff:fed8:3139"],"agent.type":["filebeat"],"host.os.kernel":["5.13.0-52-generic"],"fleet.migration.deleted":[0],"fleet.migration.updated":[1],"elastic_agent.snapshot":[true],"host.id":["c287e219ca4f4643aa206fc914242e26"],"fleet.migration.es.took":[26],"service.name":["fleet-server"],"elastic_agent.id":["308052ab-bca4-4eaa-8787-904feea56fa0"],"data_stream.namespace":["default"],"host.os.codename":["impish"],"message":["migration AgentOutputs done"],"event.ingested":["2022-08-12T10:00:51.000Z"],"@timestamp":["2022-08-12T10:00:43.961Z"],"host.os.platform":["ubuntu"],"data_stream.dataset":["elastic_agent.fleet_server"],"log.file.path":["/opt/Elastic/Agent/data/elastic-agent-6d830e/logs/default/fleet-server-20220812-15.ndjson"],"fleet.migration.batches":[1],"agent.ephemeral_id":["b21f3b26-c6a3-4da6-9edf-56fc01813ebf"],"event.dataset":["elastic_agent.fleet_server"]}}
```

## Related issues

- Fixes #1672
